src/main/scala/uk/gov/nationalarchives/omega/api/services/Dispatcher.scala

```diff
@@ -0,0 +1,107 @@
+
+import cats.data.Validated.{ Invalid, Valid }
+import cats.data.{ Validated, ValidatedNec }
+import cats.effect.IO
+import cats.effect.std.Queue
+import uk.gov.nationalarchives.omega.api.business._
+import uk.gov.nationalarchives.omega.api.business.echo.{ EchoRequest, EchoResponse, EchoService }
+import uk.gov.nationalarchives.omega.api.services.ServiceIdentifier.ECHO001
+
+import java.util.UUID
+
+class Dispatcher(val localProducer: LocalProducer) {
+
+  def run(dispatcherId: Int)(q: Queue[IO, LocalMessage]): IO[Unit] =
+    for {
+      requestMessage <- q.take
+      _ <- IO.delay {
+             println(s"Dispatcher # $dispatcherId, processing message id: ${requestMessage.persistentMessageId}")
+           }
+      serviceRequest                  <- createServiceRequest(requestMessage)
+      businessService                 <- IO.pure(serviceRequest._1)
+      businessServiceRequest          <- IO.pure(serviceRequest._2)
+      validatedBusinessServiceRequest <- validateBusinessServiceRequest(businessService, businessServiceRequest)
+      businessResult                  <- execBusinessService(businessService, validatedBusinessServiceRequest)
+      res                             <- sendResultToJmsQueue(businessResult, requestMessage)
+    } yield res
+
+  private def createServiceRequest(localMessage: LocalMessage): IO[(BusinessService, BusinessServiceRequest)] =
+    localMessage.serviceId match {
+      case ECHO001 =>
+        IO.delay {
+          Tuple2(
+            new EchoService(),
+            EchoRequest(localMessage.messageText)
+          )
+        }
+      // add more service IDs here
+    }
+
+  private def validateBusinessServiceRequest(
+    businessService: BusinessService,
+    businessServiceRequest: BusinessServiceRequest
+  ): IO[ValidatedNec[RequestValidationError, BusinessServiceRequest]] =
+    IO.delay {
+      businessService match {
+        case value: RequestValidation =>
+          value.validateRequest(businessServiceRequest)
+        case _ =>
+          Validated.valid(businessServiceRequest)
+      }
+    }
+
+  private def execBusinessService[T <: BusinessServiceRequest, U <: BusinessServiceResponse, E <: BusinessServiceError](
+    businessService: BusinessService,
+    validatedBusinessServiceRequest: ValidatedNec[RequestValidationError, T]
+  ): IO[ValidatedNec[RequestValidationError, Either[BusinessServiceError, BusinessServiceResponse]]] =
+    IO.delay {
+      validatedBusinessServiceRequest.map(businessService.process)
+    }
+
+  private def sendResultToJmsQueue[U <: BusinessServiceResponse, E <: BusinessServiceError](
+    businessResult: ValidatedNec[RequestValidationError, Either[E, U]],
+    requestMessage: LocalMessage
+  ): IO[Unit] = {
+    val replyMessage: IO[String] = IO.delay {
+      businessResult match {
+        case Valid(Right(businessResult)) =>
+          requestMessage.serviceId match {
+            case ECHO001 => businessResult.asInstanceOf[EchoResponse].text
```

Comment:
I agree and I've removed it by putting the text (or rather content) into the trait. However I think the requirements are a little hazy here until we add a meaningful service, so I expect this will change.